### DISCUSSION TITLE
Is there a limit of requests per  connection?

### DISCUSSION THREAD

**Discussion by atimin** [tokio-rs/axum#2033] [Q&A] [answered]
Hey all,

I'm using `axum` as an HTTP front-end for my database. Very often, it has to receive thousands of requests from a client. After some number of requests, the HTTP server drops the connection, and I get a `ServerDisconnect` error. I use the Python `aiohttp` library, and people suggest a workaround:

https://stackoverflow.com/a/67641667/1077167

For me, it also works to send "Connection: close" every 1000th request to renew the current connection.

I've tried to find something about the limit for HTTP requests in Hyper and Axum documentation but didn't find any mention. I also tried to change the limit by using `Keep-Alive: timeout=Y max=X` header, but it looks like it doesn't change the behavior.

Could someone explain how Axum or Hyper manages this limit?

Thanks!

**Comment by davidpdrsn** [comment, marked as answer]
I'm not aware of such a limit. Feels like it might be the client's fault. Are you able to reproduce it with something like [reqwest](https://crates.io/crates/reqwest)?

You can also try asking in `#hyper` in the [tokio discord](https://discord.gg/tokio). axum doesn't itself deal with connections directly, that's all delegated to hyper.

**Reply by atimin** [reply]
Thanks for the response. I've written a Rust program by using reqwest to repeat the issue. You're right, it is something in my Python client , because the Rust program sends thousands requests without any problems.

However, if the server returns an HTTP error for a POST request, `reqwest` crashes with a `Broken pipe` even if I use the "Expect: 100-continue" header. I' wrote about it here https://github.com/tokio-rs/axum/discussions/2016. I'm getting more confident that this case doesn't handle properly on the server side. I'll follow to your advice and write about it in the tokio community.

**Reply by atimin** [reply]
Sorry, I wasn't precise. It is not `Broken pipe` in the Rust client, it is `IncompleteMessage`:
```
Error: reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("play.reduct.store")), port: None, path: "/api/v1/b/datasets/imdb", query: Some("q=236"), fragment: None }, source: hyper::Error(IncompleteMessage) }
```
